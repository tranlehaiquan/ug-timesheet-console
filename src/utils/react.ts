import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
