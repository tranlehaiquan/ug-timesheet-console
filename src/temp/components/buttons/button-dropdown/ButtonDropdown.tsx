import * as React from 'react'

import { Button, ButtonTypes } from '../button/Button'
import { Icon } from '../../icon/Icon'
import { Dropdown } from '../../dropdown/Dropdown'

interface IButtonDropdownProps extends React.HTMLAttributes<HTMLButtonElement> {
  /**
   * Text to display on the button
   */
  label: string
  /**
   * Pass additional class names
   */
  className?: string
  /**
   * Compact sized buttonâ€”used when space is tight - allowing more room for content
   */
  compact?: boolean
  /**
   * Disable the button
   */
  disabled?: boolean
  /**
   * Button type styles
  */
  buttonType?: ButtonTypes
}

export interface IButtonDropdownState {
  activeDropdown: boolean
}

/**
 * The DropdownButton component displays/hides children as a dropdown when clicked. The Menu and MenuItem
 components are strongly recommended to be used in conjuction with this component to enforce correct styling of the dropdown menu.
 */
export class ButtonDropdown extends React.PureComponent<IButtonDropdownProps, IButtonDropdownState> {
  constructor(props: IButtonDropdownProps) {
    super(props)

    this.state = {
      activeDropdown: false
    }
  }

  componentDidMount() {
    window.addEventListener('resize', this.hideDropdown)
  }

  componentWillUnmount() {
    this.closeDropdown()
    window.removeEventListener('resize', this.hideDropdown)
  }

  closeDropdown = () => {
    if (this.state.activeDropdown) {
      this.hideDropdown()
      document.removeEventListener('click', this.closeDropdown)
    }
  }

  listenerClick = (e: any) => {
    this.showDropdown()
    document.addEventListener('click', this.closeDropdown)
  }

  showDropdown = () => {
    this.setState({ activeDropdown: true })
  }

  hideDropdown = () => {
    this.setState({ activeDropdown: false })
  }

  render() {
    const { compact, label, onClick, children, className, ...otherProps } = this.props
    const { activeDropdown } = this.state

    return (
      <Dropdown
        visible={ activeDropdown }
        placement="bottom-start"
        trigger={
          <Button
            className={ className }
            buttonType={ this.props.buttonType || 'secondary' }
            compact={ compact }
            onClick={ this.listenerClick }
            { ...otherProps }
          >
            { label }
            <Icon name="chevronDown" className="sk-ml-2" size={ 8 } />
          </Button>
        }
      >
        { children }
      </Dropdown>
    )
  }
}
