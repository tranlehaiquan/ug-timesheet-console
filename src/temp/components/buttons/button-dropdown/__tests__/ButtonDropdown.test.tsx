import * as React from "react";
import { render, shallow, mount } from "enzyme";

import { ButtonDropdown } from "../ButtonDropdown";

describe("ButtonDropdown", () => {
  test("it renders correctly", () => {
    // Arrange
    const component = render(<ButtonDropdown label="Button dropdown" />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("it renders with compact prop", () => {
    // Arrange
    const component = render(
      <ButtonDropdown label="Button dropdown" compact />
    );

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("renders with correct initial state", () => {
    // Arrange
    const component = shallow<ButtonDropdown>(
      <ButtonDropdown label="Button dropwdown" />
    );

    // Assert
    expect(component.state().activeDropdown).toBe(false);
  });

  test("display dropdown when button clicked", () => {
    // Arrange
    const component = mount<ButtonDropdown>(
      <ButtonDropdown label="Button dropdown">
        <div data-sk-name="child-dropdown-item">child dropdown menu</div>
      </ButtonDropdown>
    );

    // Act
    expect(component.find('[data-sk-name="child-dropdown-item"]')).toHaveLength(
      0
    );
    component.find("button").simulate("click");
    expect(component.find('[data-sk-name="child-dropdown-item"]')).toHaveLength(
      1
    );
  });
});
