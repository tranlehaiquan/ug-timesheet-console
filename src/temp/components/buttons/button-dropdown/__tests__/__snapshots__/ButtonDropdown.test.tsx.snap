// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonDropdown it renders correctly 1`] = `
<div
  data-sk-name="dropdown-reference"
>
  <button
    class="sked-button sked-button--secondary"
  >
    <div
      class="sk-visible"
    >
      Button dropdown
      <svg
        class="sk-flex-no-shrink sk-fill-current sk-ml-2"
        height="8"
        viewBox="0 0 18 11.17"
        width="8"
      >
        <path
          d="M.62.63A2.18 2.18 0 0 0 .7 3.75l6.71 6.75a2.23 2.23 0 0 0 3.18 0l6.71-6.75a2.18 2.18 0 0 0 .08-3.12 2.16 2.16 0 0 0-3.11.07L9 6 3.73.7A2.15 2.15 0 0 0 .62.63z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
    >
      <div
        class="loading-spinner sk-invisible"
      >
        <svg
          class="loading-spinner--circular"
          height="18"
          viewBox="25 25 50 50"
          width="18"
        >
          <circle
            class="loading-spinner--path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke="currentColor"
            stroke-miterlimit="10"
            stroke-width="5"
          />
        </svg>
      </div>
    </div>
  </button>
</div>
`;

exports[`ButtonDropdown it renders with compact prop 1`] = `
<div
  data-sk-name="dropdown-reference"
>
  <button
    class="sked-button sked-button--secondary sked-button--compact"
  >
    <div
      class="sk-visible"
    >
      Button dropdown
      <svg
        class="sk-flex-no-shrink sk-fill-current sk-ml-2"
        height="8"
        viewBox="0 0 18 11.17"
        width="8"
      >
        <path
          d="M.62.63A2.18 2.18 0 0 0 .7 3.75l6.71 6.75a2.23 2.23 0 0 0 3.18 0l6.71-6.75a2.18 2.18 0 0 0 .08-3.12 2.16 2.16 0 0 0-3.11.07L9 6 3.73.7A2.15 2.15 0 0 0 .62.63z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
    >
      <div
        class="loading-spinner sk-invisible"
      >
        <svg
          class="loading-spinner--circular"
          height="18"
          viewBox="25 25 50 50"
          width="18"
        >
          <circle
            class="loading-spinner--path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke="currentColor"
            stroke-miterlimit="10"
            stroke-width="5"
          />
        </svg>
      </div>
    </div>
  </button>
</div>
`;
